src/types.md

```diff
@@ -175,4 +175,47 @@ for example.
 To generalize this idea, try to avoid storing references to anything that might
 need to be changed. Instead take those things as parameters. For instance
 `petgraph` [takes the entire graph as context to a `Walker` object](https://docs.rs/petgraph/0.6.0/petgraph/visit/trait.Walker.html),
-such that the graph can be changed while you're walking it.
\ No newline at end of file
+such that the graph can be changed while you're walking it.
+
+## My type needs to store arbitrary user data. What do I do instead of `void *`?
+
+Ideally, your type would know all possible types of user data that it would store.
+You'd encode this as an `enum` with variant data for each possibility. This would
+give complete compile-time type safety.
+
+But there may be some patterns where one area of code needs to store data that's
+defined by a totally different area of the codebase, or belongs to clients, and
+such possibilities can't be enumerated in advance. In such cases in C++ you'd
+use a `void *` and clients would downcast to get their original type back.
+
+To do this in Rust, define a `trait` (let's call it `UserData`) and then store a
+`Box<dyn UserData>`.
```

Comment:
I found this confusing because it seems to imply that you need to define your own trait to use `Any`.

Maybe start with a simpler example first that introduces `Any` -- something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fd0b3c4aa55cfbfe22807472593d0413)?

And then progress to discussing how to use `Any` with traits -- the `as_any` pattern certainly does seem important enough to discuss.